File: yujung/PRG/datastructure/올바른괄호.py
def solution(s):
    
    answer = True
    
    parenthesis = list(s) # 문자열를 구성하는 문자들을 하나하나 리스트에 담는다
    
    stack = []
    for i in parenthesis:
        if i == '(':
            stack.append('(')
        elif i == ')':
            if len(stack) == 0: # ')' 문자가 등장하기 이전에 스택에 '('가 존재해야 한다
                return False
            else:
                stack.pop() # ')' 문자가 등장하면 스택에 쌓인 '(' 문자를 스택에서 뽑아낸다
    
    if len(stack) != 0: # 모든 괄호가 짝이 맞은 경우 스택이 비어 있어야 한다
        return False

    return True

File: NTY/PRG/자료구조/올바른괄호/올바른괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        Stack<Character> stack = new Stack<>();
        char[] cs = s.toCharArray();
        for(char c : cs){
            if(c == '(') stack.push(c);
            else if(c == ')'){
                if(!stack.empty() && stack.peek() == '(') stack.pop();
                else return false;
            }
        }
        return stack.empty();
    }
}